### PR TITLE
Fixed randomness in tl.diffmap - compute_eigen (v0 argument)

### PR DESCRIPTION
**Pull request by Iwo-K**
Currently tl.diffmap generates slightly different results on consequent runs. This is because the v0 argument is not provided in the scipy.sparse.linalg.eigsh call (inside the compute_eigen function). 

To fix this I set the random_state inside the compute_eigen function and generate a random vector to pass into scipy.sparse.linalg.eigsh as v0. 
I tried to follow how random_state is implemented in other scanpy functions, I hope this is consistent.

To allow user control of the random_state I added random_state arguments to:
- tools/_diffmap.py _diffmap function (tl.diffmap function)
- tools/_dpt.py _diffmap function
- neighbors/__init__.py compute_eigen function

### CODE REVIEW

**Review by ivirshup — CHANGES_REQUESTED** [review]
Thanks for the PR!

I think the way we're supposed to be using random state internally (due to recent numpy updates) looks more like:

```python
random_state = check_random_state(random_state)
v0 = random_state.standard_normal(matrix.shape[0])
```

However, I recall there being some weirdness about scipy arpack and random state where it was difficult to ever get a truly deterministic result (see #1749, https://github.com/scipy/scipy/issues/10374).

Could add a test to make sure this works? Something checking that case which is currently non-deterministic becomes deterministic after your patch.

**Comment by Iwo-K** [comment]
Sure, will do! Should I combine this PR with the fix mentioned in issue #313 or would you prefer a separate PR?

**Comment by ivirshup** [comment]
Thanks! Separate PR for that, please!

**Comment by Iwo-K** [comment]
Done! I've changed the random vector generation to use the code you suggested and added a test to test_embedding.

File: scanpy/neighbors/__init__.py [hunk 1]
```diff
@@ -978,8 +981,13 @@ def compute_eigen(
             which = 'LM' if sort == 'decrease' else 'SM'
             # it pays off to increase the stability with a bit more precision
             matrix = matrix.astype(np.float64)
+
+            # Setting the random initial vector
+            random_state = check_random_state(random_state)
+            np.random.set_state(random_state.get_state())
```
Comment (ivirshup) [inline scanpy/neighbors/__init__.py]: Is this line necessary for reproducibility? If not, I don't think we have to modify the global state.

File: scanpy/tools/_diffmap.py [hunk 2]
```diff
@@ -39,6 +41,8 @@ def diffmap(
         .obsp[.uns[neighbors_key]['connectivities_key']],
         .obsp[.uns[neighbors_key]['distances_key']] for connectivities and distances
         respectively.
+    random_state
+            A numpy random seed
```
Comment (ivirshup) [inline scanpy/tools/_diffmap.py]:
```suggestion
        A numpy random seed
```

**Review by ivirshup — COMMENTED** [review]
Added some minor comments (one formatting, possibly one unnecessary line), but looks great overall.

Could you add a bug-fix release note to `docs/release-notes/1.8.0.rst`?

File: scanpy/neighbors/__init__.py (hunk 1 above)
Comment (Iwo-K) [reply scanpy/neighbors/__init__.py]: I've just checked, specifying the v0 is enough to get the same results every time. No need for setting global state. I will delete the line, thanks!

**Comment by Iwo-K** [comment]
No problem, all done!

**Comment by ivirshup** [comment]
Thanks for the PR!

**Pull request merged**